### DISCUSSION TITLE
Simple Light Client

### DISCUSSION THREAD

**Discussion by Shadow8472** [smcameron/space-nerds-in-space#334] [Ideas]
Considering my diagnostics in #333 

My aged computer ran Space Nerds decently well on reduced textures, but suppose things were taken farther for computers with even less business running main screen/weapons. My suggestion is to modify the launcher to offer starting in a no-textures mode as opposed to insisting on downloading assets. Minimal models for making the screen and wireframe workstations are then loaded and everything else is either touched or linked to the same error texture.

**Comment by smcameron** [comment]
Hmm.  Not really sure I'm all that interested in making the game that ugly, lol.  And I don't know that a single error texture would work.  For example, if the skybox and the planets used the same default texture, e.g. a black square, the planets would be invisible.  If the skybox used a white square, then the sky would look awful.  Partially transparent textures are another case that would look awful with default textures (nebulas, thruster exhaust, laser beams, torpedoes, planetary rings, explosion-sparks, lens flare, etc.)

There are old videos before I used opengl... back when the game was kind of a 2D game that happened to have a 3D renderer... https://www.youtube.com/watch?v=KcEbCyKU0Qs

Not sure I want to go back to that look.

**Reply by Shadow8472** [reply]
My proposed implementation by faking out the textures was meant to minimize the changes to the client. I suppose if the main menu were aware of this Light Client Mode the Main Screen/Weapons roles could be greyed out and ask for full assets before unlocking. 

No outside view means no need for outside textures and greater accessibility to groups with only a couple respectable GPU's between them (speaking from my experience before your script to scale down Zolon). Everyone else just gets enough to run the other stations.

**Comment by smcameron** [comment]
Ah, ok, I see.  Yeah, some of the stations ought to be able to run without textures.  I'm a little busy this week, but I'll think about this, maybe I can get to it next week.

**Comment by smcameron** [comment]
![notextures](https://github.com/smcameron/space-nerds-in-space/assets/23912/2210f141-9889-4544-8b17-cd3ac17f4d81)
Committed f0726df283fe765c15bb1ee80ac8de6841b21b9b snis_client: add no-texture mode

You can pass "--no-textures" to snis_client, or via snis_launcher, in the options menu there's an item to turn on the --no-textures flag.

It does still load one small 128x128 texture that's a green grid on a black background that it uses for everything.

**Comment by smcameron** [comment]
Taking a quick look at RESident memory usage with htop, using the --no-textures option drops the memory usage of snis_client from 699M to 295M on my system.

**Reply by Shadow8472** [reply]
Impressive. I suppose that the choke point in #333 lowered.

**Reply by smcameron** [reply]
> Impressive. I suppose that the choke point in #333 lowered.

A few hundred Mb here, a few hundred there, and pretty soon you're talking real memory.

**Reply by Shadow8472** [reply]
But it was GPU memory I was experiencing the bottleneck with.

I took a look at the changelog. Does starting in --no-textures mode still insist update-assets be run?

**Comment by smcameron** [comment]
It does still load *a* texture to the GPU, the same 128x128 texture over and over.  It could be made even smaller.  There might be a further optimization to load the same texture once, and use the same texture ID everywhere, but I'm not sure how practical that is.  The change I made is sort of the smallest change I could make that moves in the direction of not needing textures.  It should load the same 128x128 texture 135 times by my count, counted via this patch.  I expect it should take about 8Mb of GPU memory, plus space for mipmaps, so probably around 10-ish Mb.  You could reduce it further by scaling small-green-grid.png down to 64x64 or 32x32 or 16x16 or maybe even down to 1x1, then it should be down into "few kilobytes" territory.

```
diff --git a/png_utils.c b/png_utils.c
index 2a2f1c41..9ccb3ebd 100644
--- a/png_utils.c
+++ b/png_utils.c
@@ -106,6 +106,7 @@ char *png_utils_read_png_image(const char *filename, int flipVertical, int flipH
        int *w, int *h, int *hasAlpha, char *whynot, int whynotlen)
 {
        int i, j, bit_depth, color_type, row_bytes, image_data_row_bytes;
+       static int count = 0;
        png_byte header[8];
        png_uint_32 tw, th;
        png_structp png_ptr = NULL;
@@ -114,6 +115,8 @@ char *png_utils_read_png_image(const char *filename, int flipVertical, int flipH
        png_byte *image_data = NULL;
        png_byte **image_data_ptr = NULL;
 
+       printf("Loading png image %s (%d)\n", filename, count++);
+
        FILE *fp = fopen(filename, "rb");
        if (!fp) {
                snprintf(whynot, whynotlen, "Failed to open '%s': %s",
```

You should not have to run "make update-assets". The only PNG files it should look for are

```
./share/snis/textures/snis-splash.png
./share/snis/textures/small-green-grid.png
```
And if it doesn't find the first one, it should continue anyway, because that's for the splash screen at the beginning which runs in a separate process and so it will either proceed without the image, or die, and it doesn't matter which happens, snis_client doesn't care if the splash screen process succeeds or fails.

In theory, you don't *ever* have to run "make update-assets", there are supposed to be enough assets already in the git repo to get by, and "make update-assets" just downloads more and better assets.  However, I have not run without doing "make update-assets" in a long long time, so it is quite possible, maybe even likely, that something is broken if you try skipping "make update-assets", but in theory it should work even without --no-textures, but it definitely should be ok to skip "make update-assets" with --no-textures.

**Reply by Shadow8472** [reply]
Remembering the splash screen, in my quest for a low-memory SNIS option, I tried and failed spectacularly (machine froze) to load SNIS in the Wayland compositor, Cage. It hung for a while and crashed -- presumably just long enough for the launcher to try its thing with nothing to show but a mouse pointer. Easily a case of PEBKaC, as I couldn't get keyboard/mouse to behave with SuperTuxKart either, though STK did render (though on further inspection, it appears to have had Wayland support in the works at some time).

**Reply by smcameron** [reply]
FWIW, I believe SNIS should work with Wayland, although I haven't tried it myself.  Someone in this bug report got it working: https://github.com/smcameron/space-nerds-in-space/issues/301

**Comment by smcameron** [comment]
Ok, I just tried a fresh clone of space-nerds-in-space and tried to run it without "make update-assets".

```
Failed to find ./share/snis/models/cylinder.stl
Perhaps you need to run 'make update-assets' or 'make assets'?
```

Oh yeah...  Only the OpenSCAD models are in git, you still need the stl files to be generated...

Hmm, "make assets" doesn't work (anymore?)

Alright need to dig into this.

**Comment by smcameron** [comment]
Oh, it's "make models", not "make assets".

And this will run OpenSCAD a bunch of times on all the share/snis/models/\*.scad file to produce all the share/snis/models/\*.stl files, which takes forever (even worse on an old slow machine.)

But, it does work, tried it both with and without --no-textures.

**Comment by smcameron** [comment]
Committed the following:

* 2d75a95f67a515546530498f0358e2b59ff5f704 snis_launcher: correct error in instructions for "make models"
